crates/assists/src/handlers/inline_function.rs

```diff
@@ -0,0 +1,229 @@
+            inline_function,
+            r#"
+            fn foo() { println!("Hello, World!"); }
+            fn main() {
+                fo<|>o();
+            }
+            "#,
+            r#"
+            fn foo() { println!("Hello, World!"); }
+            fn main() {
+                {
+                    println!("Hello, World!");
+                };
+            }
+            "#,
+        );
+    }
+
+    #[test]
+    fn args_with_side_effects() {
+        check_assist(
+            inline_function,
+            r#"
+            fn foo(name: String) { println!("Hello, {}!", name); }
+            fn main() {
+                foo<|>(String::from("Michael"));
+            }
+            "#,
+            r#"
+            fn foo(name: String) { println!("Hello, {}!", name); }
+            fn main() {
+                {
+                    let name = String::from("Michael");
+                    println!("Hello, {}!", name);
+                };
+            }
+            "#,
+        );
+    }
+
+    #[test]
+    #[ignore = "not implemented"]
+    fn copy_args_with_no_side_effects_get_inlined() {
+        check_assist(
+            inline_function,
+            r#"
+            fn add(a: u32, b: u32) -> u32 { a + b }
+            fn main() {
+                let x = add<|>(1, 2);
+            }
+            "#,
+            r#"
+            fn add(a: u32, b: u32) -> u32 { a + b }
+            fn main() {
+                let x = 1 + 2;
+            }
+            "#,
+        );
+    }
+
+    #[test]
+    fn cant_inline_when_the_function_is_inaccessible() {
+        check_assist_not_applicable(
+            inline_function,
+            r#"
+            mod private {
+                fn add(a: u32, b: u32) -> u32 { a + b }
+            }
+            fn main() { let x = private::add<|>(1, 2); }
+            "#,
+        );
+    }
+
+    #[test]
+    fn method_inlining_isnt_supported() {
+        check_assist_not_applicable(
+            inline_function,
+            r"
+            struct Foo;
+            impl Foo { fn bar(&self) {} }
+
+            fn main() { Foo.bar<|>(); }
+            ",
+        );
+    }
+
+    #[test]
+    fn function_with_multiple_statements() {
+        check_assist(
+            inline_function,
+            r#"
+            fn foo(a: u32, b: u32) -> u32 {
+                let x = a + b;
+                let y = x - b;
+                x * y
+            }
+
+            fn main() {
+                let x = foo<|>(1, 2);
+            }
+            "#,
+            r#"
+            fn foo(a: u32, b: u32) -> u32 {
+                let x = a + b;
+                let y = x - b;
+                x * y
+            }
+
+            fn main() {
+                let x = {
+                    let a = 1;
+                    let b = 2;
+                    let x = a + b;
+                    let y = x - b;
+                    x * y
+                };
+            }
+            "#,
+        );
+    }
+}
```

Comment:
We try to use a slightly different indent style: https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/style.md#minimal-tests